server/service/osquery.go

```diff
@@ -1405,3 +1414,155 @@ func (svc *Service) SubmitResultLogs(ctx context.Context, logs []json.RawMessage
 	}
 	return nil
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Query Reports
+////////////////////////////////////////////////////////////////////////////////
+
+func (svc *Service) SaveResultLogsToQueryReports(ctx context.Context, results []json.RawMessage) error {
+	// skipauth: Authorization is currently for user endpoints only.
+	svc.authz.SkipAuthorization(ctx)
+
+	// Do not insert results if query reports are disabled globally
+	appConfig, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		return newOsqueryError("getting app config: " + err.Error())
+	}
+	if appConfig.ServerSettings.QueryReportsDisabled {
+		return nil
+	}
+
+	var queryResults []fleet.ScheduledQueryResult
+	for _, raw := range results {
+		var result fleet.ScheduledQueryResult
+		if err := json.Unmarshal(raw, &result); err != nil {
+			return newOsqueryError("unmarshalling individual result log: " + err.Error())
+		}
+		queryResults = append(queryResults, result)
+	}
+
+	// Filter results to only the most recent for each query
+	filtered := getMostRecentResults(queryResults)
+
+	for _, result := range filtered {
+		if err := svc.processResults(ctx, result); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (svc *Service) processResults(ctx context.Context, result fleet.ScheduledQueryResult) error {
+	// Discard result if there is no snapshot
+	if len(result.Snapshot) == 0 {
+		return nil
+	}
+
+	teamID, queryName, err := getQueryNameAndTeamIDFromResult(result.QueryName)
+	if err != nil {
+		return newOsqueryError("getting query name and team ID: " + err.Error())
+	}
+
+	query, err := svc.ds.QueryByName(ctx, teamID, queryName)
+	if err != nil {
+		return newOsqueryError("getting query by Name: " + err.Error())
+	}
+
+	// Discard Result if query is marked as discard data or if the query report is full
+	if query.DiscardData {
+		return nil
+	}
```

Comment:
Should we check if `query.Logging` is `"snapshot"`? (You could be getting results from when it was configured to run as snapshot, but then the user might have changed to differential recently.)

Reply:
good catch, i have that handled during ingest if `len(snapshot) == 0` but we should check in the policy settings as well